src/orange/orangeinp/ConvexHullFinder.hh

```diff
@@ -0,0 +1,289 @@
+
+            concave_region.push_back(points_[i]);
+            concave_regions.push_back(std::move(concave_region));
+        }
+    }
+    return concave_regions;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate a mask that indicates which points are on the convex hull.
+ *
+ * This method uses the Graham Scan algorithm.
+ */
+template<class T>
+auto ConvexHullFinder<T>::calc_convex_mask() const -> ConvexMask
+{
+    // Find the indices of the points on the convex hull. Start from the point
+    // with the lowest y value, which is guaranteed to be on the hull.
+    std::vector<size_type> hull;
+    auto i = start_index_;
+    hull.push_back(i);
+    i = this->calc_next(i);
+
+    for (size_type j = 0, n = points_.size(); j != n; ++j)
+    {
+        size_type i_next = this->calc_next(i);
+
+        if (this->is_clockwise(hull.back(), i, i_next))
+        {
+            // Clockwise point is part of the hull
+            hull.push_back(i);
+        }
+        else
+        {
+            // Pop points off the hull until we can reach the next point by
+            // turning clockwise.
+            while (hull.size() >= 2
+                   && !this->is_clockwise(
+                       hull[hull.size() - 2], hull.back(), i_next))
+            {
+                hull.pop_back();
+            }
+        }
+
+        i = i_next;
+    }
+
+    // Convert convex hull indices to a mask
+    ConvexMask convex_mask(points_.size(), false);
+    for (auto h : hull)
+    {
+        convex_mask[h] = true;
+    }
+
+    return convex_mask;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find the index of the element with the lowest y value.
+ */
+template<class T>
+size_type ConvexHullFinder<T>::min_element_idx() const
+{
+    auto starting_it = std::min_element(
+        points_.begin(), points_.end(), [](Real2 const& a, Real2 const& b) {
+            return a[1] < b[1];
+        });
+    return std::distance(points_.begin(), starting_it);
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine if three elements form a clockwise turn using the cross product.
+ *
+ * Here, colinear points are considered clockwise.
+ */
+template<class T>
+auto ConvexHullFinder<T>::is_clockwise(size_type i_prev,
+                                       size_type i,
+                                       size_type i_next) const -> bool
+{
+    auto const& a = points_[i_prev];
+    auto const& b = points_[i];
+    auto const& c = points_[i_next];
+
+    // Create a SoftOrientation object based on the bump distance of the middle
+    // point. Since the BumpCalculator operates on Real3, use a z-value of 0.
+    auto abs_tol = calc_bump_({b[0], b[1], 0});
+    detail::SoftOrientation<real_type> soft_ori_(abs_tol);
```

Comment:
Is it reasonable to precalculate this based on the extents of the whole polygon? It seems like the tolerance should be relative to that, not the points being compared.

Reply:
Yes, I think that would be better. `BumpCalculator::operator()` takes a point as argument though, so I wasn't sure what you were getting at. I could make the point `(max_x - min_x, max_y - min_y, 0)` if this doesn't feel like too much of a hack.